s2a/src/main/java/io/grpc/s2a/internal/handshaker/SslContextFactory.java

```diff
@@ -138,9 +138,14 @@ private static void configureSslContextWithClientTlsConfiguration(
           NoSuchAlgorithmException,
           UnrecoverableKeyException {
     sslContextBuilder.keyManager(createKeylessManager(clientTlsConfiguration));
-    ImmutableSet<String> tlsVersions =
-        ProtoUtil.buildTlsProtocolVersionSet(
-            clientTlsConfiguration.getMinTlsVersion(), clientTlsConfiguration.getMaxTlsVersion());
+    ImmutableSet<String> tlsVersions;
+    try {
+      tlsVersions =
+          ProtoUtil.buildTlsProtocolVersionSet(
```

Comment:
When a client asks S2A for a TLS config (in Java, this is used to build the SslContext), S2A includes the [min and max TLS version](https://github.com/grpc/grpc-java/blob/master/s2a/src/main/proto/grpc/gcp/s2a/s2a.proto#L109) that the client should use, which is used to populate the SslContextBuilder.protocols. So In the case of a new TLS version, S2A would update max version in the response. 

Should the client ignore any unknown version numbers and only fail if it knows none of the versions?
> I believe this is what happens. Although I think the implementation is outside of S2A client. S2A client builds an SslContext with the response from S2A.  If the client doesn't support a particular / any of the versions specified in SslContextBuilder.protocols, the handling for that is outside S2A client. 

I'll ask Matt to add to / correct.

Reply:
Thanks for raising this point @ejona86! As @rmehta19 mentioned, the S2A sends a range of allowed TLS versions, in the form of a min/max TLS version. If one end of the range is not supported (e.g. the max is TLS 1.4), then I think we should use the intersection of the given range with the supported range. For example, if the S2A says use min=1.2 and max=1.4 and the client only supports 1.2 and 1.3, then we use 1.2 and 1.3. Does that sound reasonable?

Reply:
That sounds like reasonable behavior (to only fail when the intersection is empty).

But that is not what the code is doing. The code right here is clearly ignoring everything on IllegalArgumentException, which is thrown for an unknown TLS version.

Reply:
Yes agreed, the code needs to be updated.

Reply:
I have updated the code to not throw an error on unknown TLS version in https://github.com/grpc/grpc-java/pull/11588/commits/dfd8036380aed6376fd31ee0539e6256df49dbec

Initially, I was missing the part where the binary could know different versions then what S2A responds with. So if we add 1.4, the binary could only know up to 1.3. Once it rebuilds, it will know about 1.4. But until then, we don't want to crash the binary. Thanks for bringing this up!